Vault secret manager

alvaroaleman opened the pull request:
Adds a vault secret manager, see the provided Readme for a detailled description of how it works. This PR only provides the backend part and no frontend yet. The first commit "Add a vaultclient" is in a dedicated PR: https://github.com/openshift/ci-tools/pull/1737

/cc @openshift/openshift-team-developer-productivity-test-platform 
Ref https://issues.redhat.com/browse/DPTP-1891

alvaroaleman commented:
/retest

stevekuznetsov commented on cmd/secret-collection-manager/main.go:
```diff
@@ -0,0 +1,415 @@
+	flag.StringVar(&o.kvStorePrefix, "kv-store-prefix", "secret/self-managed", "Vault KV folder under which all policies will get created")
+	flag.StringVar(&o.listenAddr, "listen-addr", "127.0.0.1:8080", "The address to listen on")
+	flag.StringVar(&o.vaultAddr, "vault-addr", "http://127.0.0.1:8300", "The address under which vault should be reached")
+	flag.StringVar(&o.vaultToken, "vault-token", "jpuxZFWWFW7vM882GGX2aWOE", "The privileged token to use when communicating with vault, must be able to CRUD policies")
```
Kind of weird that this has a default

stevekuznetsov commented on cmd/secret-collection-manager/main.go:
```diff
@@ -0,0 +1,415 @@
+		logrus.WithError(err).Fatal("Failed to construct vault client")
+	}
+
+	if err := server(privilegedVaultClient, o.kvStorePrefix, o.listenAddr).ListenAndServe(); err != nil && err != http.ErrServerClosed {
```
Might as well wrap with `interrupts.ListenAndServe`?

alvaroaleman replied:
I would be less hesitant to use `interrupts` if https://github.com/kubernetes/test-infra/issues/17737 was fixed and it wasn't entirely based on package-level variables :/

stevekuznetsov commented on cmd/secret-collection-manager/main.go:
```diff
@@ -0,0 +1,415 @@
+
+}
+
+func server(privilegedVaultClient *vaultclient.VaultClient, kvStorePrefix, listenAddr string) *http.Server {
```
What's the use of breaking this func out?

alvaroaleman replied:
That func is what the tests use

stevekuznetsov commented on cmd/secret-collection-manager/main.go:
```diff
@@ -0,0 +1,415 @@
+
+func (m *secretCollectionManager) mux() *httprouter.Router {
+	router := httprouter.New()
+	router.GET("/healthz", healthHandler)
```
We have health & readiness lib already from Prow, prefer that for consistency of code, ports, etc

alvaroaleman replied:
It doesn't really do anything different other than hardcoding a different port

stevekuznetsov commented on cmd/secret-collection-manager/main.go:
```diff
@@ -0,0 +1,415 @@
+	if err != nil {
+		return fmt.Errorf("failed to get user %s: %w", userName, err)
+	}
+	kvPathSplit := strings.Split(m.kvStorePrefix, "/")
```
Can we unit test string manipulations like this block?

stevekuznetsov commented on cmd/secret-collection-manager/main_test.go:
```diff
@@ -0,0 +1,371 @@
+		}
+	}()
+
+	if err := waitForServiceReady(ctx, t, fmt.Sprintf("http://%s/healthz", managerListenAddr)); err != nil {
```
The e2e framework has code to transparently stop, wait and stop (on timeout) processes that use the k8s ready/health endpoints (so it knows which ports to poke) - might be able to re-use?

stevekuznetsov commented on pkg/vaultclient/vaultclient.go:
```diff
@@ -0,0 +1,198 @@
+}
+
+func (v *VaultClient) readInto(path string, target interface{}) error {
+	raw, err := v.Logical().Read(path)
```
How are retries, etc handled? Inside of their client?

alvaroaleman replied:
Yeah, they use go-retriable underneath with a couple of custom settings, I thought it was best to just leave the defaults there

alvaroaleman commented:
/label tide/merge-method-squash

petr-muller commented on cmd/secret-collection-manager/main.go:
```diff
@@ -0,0 +1,419 @@
+	collections, err := m.getCollectionsForUser(l, user)
+	if err != nil {
+		l.WithError(err).Error("failed to get secret collections for user")
+		http.Error(w, fmt.Sprintf("failed to check if user is allowed to change secret collection. RequestID: %s", l.Data["UID"]), 501)
```
Curious, why 501?

petr-muller commented on cmd/secret-collection-manager/main.go:
```diff
@@ -0,0 +1,419 @@
+		return
+	}
+
+	var body secretColectionUpdateBody
```
typo: `Collection`

petr-muller commented on cmd/secret-collection-manager/main.go:
```diff
@@ -0,0 +1,419 @@
+		return fmt.Errorf("failed to serialize policy for %s: %w", secretCollectionName, err)
+	}
+	if err := m.privilegedVaultClient.Sys().PutPolicy(prefixedName(secretCollectionName), string(serializedPolicy)); err != nil {
+		println(string(serializedPolicy))
```
debugging leftover?

petr-muller commented on cmd/secret-collection-manager/main.go:
```diff
@@ -0,0 +1,419 @@
+		if err != nil {
+			return nil, fmt.Errorf("failed to resolve name for group id %s: %w", groupID, err)
+		}
+		if !strings.HasPrefix(name, objectPrefix) {
+			continue
+		}
+		groupNames = append(groupNames, name)
```
I'd consider `if HasPrefix(...) { groupNames = append(...) }` a bit more readable (getting rid of the negation and `continue`) but /shrug

petr-muller commented on cmd/secret-collection-manager/main.go:
```diff
@@ -0,0 +1,419 @@
+		if !strings.HasPrefix(path, m.kvMetadataPrefix) && !strings.HasPrefix(path, m.kvDataPrefix) {
+			return nil, fmt.Errorf("path %s in policy %s neither had the metadata(%s) nor the data(%s) prefix", path, group.Policies[0], m.kvDataPrefix, m.kvDataPrefix)
+		}
+		collection.Name = strings.TrimSuffix(strings.TrimPrefix(strings.TrimPrefix(path, m.kvMetadataPrefix+"/"), m.kvDataPrefix+"/"), "/*")
```
:exploding_head:

alvaroaleman commented:
@petr-muller Addressed all your feedback. Also renamed the thing to `vault secret collection manager`, because without the vault I can not remember the name

petr-muller reviewed (approved):
LGTM, holding if you want more reviews, it's a big thing

/hold

alvaroaleman commented:
/hold cancel

alvaroaleman commented:
/retest
